JAX time series classification

mtsokol opened the pull request:
This PR adds a tutorial with JAX time series classification task on a FordA dataset from the [UCR/UEA archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/). It follows one of the [Keras tutorials](https://keras.io/examples/timeseries/timeseries_classification_from_scratch/). It's still missing the narrative.

mtsokol commented:
It's ready for a review!

jakevdp reviewed (approved):
Looks good – it's a bit sparse on the description of what the code is doing, but we can probably address that in a later pass.

jakevdp commented:
Please resolve the conflicts, and then we can merge this. Thanks!

mtsokol commented:
> Please resolve the conflicts, and then we can merge this. Thanks!

Rebased!

I also added a couple of descriptions before some of the code cells.

pavithraes commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+    "test_accuracy": [],
+}
+```
+
````
```suggestion

We can now train the CNN model.
We'll evaluate the model’s performance on the test set after each epoch,
and print the metrics: total loss and accuracy.

```

mtsokol replied:
Done!

pavithraes commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+    train_one_epoch(epoch)
+    evaluate_model(epoch)
+```
+
````
```suggestion

Finally, let's visualize the loss and accuracy with Matplotlib.

```

pavithraes commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+but we also need to pay attention to validation accuracy so as to spot when the model starts
+overfitting.
+
+For model early stopping and selecting best model there's [Orbax](https://github.com/google/orbax)
+library which provides checkpointing and persistence utilities.
```
```suggestion
For model early stopping and selecting best model, you can check out [Orbax](https://github.com/google/orbax),
a library which provides checkpointing and persistence utilities.
```

mtsokol replied:
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+# Time series classification with JAX
+
+In this tutorial, we're going to perform time series classification with a Convolutional Neural Network.
+We're going to use FordA dataset from the [UCR archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/).
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+The problem we're facing is to assess if an engine is malfunctioning based on recorded noises it generates.
+Each sample is comprised of noise measurements across time, together with a "yes/no" label, so it's a binary classification problem.
+
+Although convolution models are mainly associated with image processing, they are useful also for time series data as they're able to extract temporal structures.
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+- [Optax](https://github.com/google-deepmind/optax) for gradient processing and optimization.
+- [Grain](https://github.com/google/grain/) will be be used to define data sources.
+- [tqdm](https://tqdm.github.io/) for a progress bar to monitor the training progress.
+
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+We're going to use FordA dataset from the [UCR archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/).
+
+The problem we're facing is to assess if an engine is malfunctioning based on recorded noises it generates.
+Each sample is comprised of noise measurements across time, together with a "yes/no" label, so it's a binary classification problem.
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+## Dataset
+
+We load dataset files into NumPy arrays, add singleton dimention to take into
+the account convolution features, and change `-1` label to `0` value:
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+```
+
+```{code-cell} ipython3
+# Here are exemplary samples from each class
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+import tqdm
+```
+
+## Dataset
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+plt.legend()
+plt.show()
+```
+
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+train_source = DataSource(x_train, y_train)
+test_source = DataSource(x_test, y_test)
+```
+
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+    shard_options=grain.NoSharding(),  # No sharding since this is a single-device setup
+    num_epochs=1,                      # Iterate over the dataset for one epoch
+)
+
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+interface. See [PyGrain Data Sources](https://github.com/google/grain/blob/main/docs/data_sources.md)
+for more details.
+
+Our dataset is comprised of relatively small NumPy arrays so our DataSource is uncomplicated:
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+)
+```
+
+## Model
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
```diff
@@ -0,0 +1,346 @@
+## Model
+
+Here we construct the model with three convolution and dense layers. We use ReLU activation
+function for middle layers and softmax in the final layer for binary classification output:
```
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+
+optimizer = nnx.Optimizer(model, optax.adam(learning_rate, momentum))
+```
+
````
Done!

mtsokol commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,346 @@
+    ).mean()
+    return loss, logits
+```
+
````
Done!

trallard commented:
@jakevdp this should be ready now 🚀

jakevdp commented on docs/JAX_time_series_classification.md:
````diff
@@ -0,0 +1,389 @@
+
+```{code-cell} ipython3
+model = MyModel(rngs=nnx.Rngs(0))
+nnx.display(model)
````
This cell is producing a GPU warning – let's remove it from the notebook output to avoid distracting readers.

mtsokol replied:
Done!

I squashed the PR, it's ready from my side!
